File: fila/Fila Estática/Java/Fila Dupla Terminação/FilaEstaticaDuplaTerminacao.java
public class FilaEstaticaDuplaTerminacao implements DuplamenteEnfileiravel {
	
	private int ponteiroInicio;
	private int ponteiroFim;
	private Object[] dados;
	private int quantidade;
	
	public FilaEstaticaDuplaTerminacao(int tamanho) {
		ponteiroInicio = 0;
		ponteiroFim = -1;
		quantidade = 0;		
		dados = new Object[tamanho];
	}

	public FilaEstaticaDuplaTerminacao() {
		this(10);
	}

	//funciona como o enfileirar de FilaEstaticaCircular
	@Override
	public void enfileirarFim(Object dado) {
		if (!estaCheia()){
			ponteiroFim = avancar(ponteiroFim);
			dados[ponteiroFim] = dado;
			//não deixar ponteiroInicio esquecido, caso a estrutura esteja na 1ª inserção
			if (estaVazia())
				ponteiroInicio = ponteiroFim;
			
			quantidade++;
		} else {
			System.err.println("Queue is full!");
		}	
	}
	
	//funciona como o desenfileirar de FilaEstaticaCircular
	@Override	
	public Object desenfileirarInicio() {
		Object dadoInicio = null;
		if (!estaVazia()){
			dadoInicio = dados[ponteiroInicio];
			ponteiroInicio = avancar(ponteiroInicio);			
			quantidade--;
		} else {
			System.err.println("Queue is empty!");
		}
		return dadoInicio;			
	}
	
	//funciona como o frente de FilaEstaticaCircular
	@Override
	public Object frente() {
		Object dadoInicio = null;
		if (!estaVazia())
			dadoInicio = dados[ponteiroInicio];
		else
			System.err.println("Fila Vazia!");		

		return dadoInicio;
	}
	
	//NOVO
	@Override
	public void enfileirarInicio(Object dado) {
		if (!estaCheia()){
			ponteiroInicio = retroceder(ponteiroInicio);
			dados[ponteiroInicio] = dado;
			//não deixar ponteiroFim esquecido, caso a estrutura esteja na 1ª inserção
			if (estaVazia())
				ponteiroFim = ponteiroInicio;
			
			quantidade++;
		} else {
			System.err.println("Queue is full!");
		}
	}
	
	//NOVO	
	@Override
	public Object desenfileirarFim() {
		Object dadoFim = null;
		if (!estaVazia()){
			dadoFim = dados[ponteiroFim];
			ponteiroFim = retroceder(ponteiroFim);
			quantidade--;
		} else {
			System.err.println("Queue is empty!");
		}
		return dadoFim;			
	}
	
	//NOVO	
	@Override
	public Object tras() {
		Object dadoFim = null;	
		if (!estaVazia()) {
			dadoFim = dados[ponteiroFim];
		} else {
			System.err.println("Fila Vazia!");		
		}
		return dadoFim;
	}

	//funciona como o atualizarInicio de FilaEstaticaCircular
	@Override
	public void atualizarInicio(Object dado) {
		if (!estaVazia()){
			dados[ponteiroInicio] = dado;
		} else {
			System.err.println("Queue is empty!");
		}
	}	

	//funciona como o atualizarFim de FilaEstaticaCircular
	@Override
	public void atualizarFim(Object dado) {
  		if (!estaVazia()){
			dados[ponteiroFim] = dado;
		} else {
			System.err.println("Queue is empty!");
		}
	}
	
	//funciona como o estaCheia de FilaEstaticaCircular	
	@Override
	public boolean estaCheia() {
		return (quantidade == dados.length);
	}

	//funciona como o estaVazia de FilaEstaticaCircular
	@Override
	public boolean estaVazia() {
		return (quantidade == 0);	
	}
	
	//funciona como o imprimir de FilaEstaticaCircular
	@Override
	public String imprimirDeFrentePraTras() {
		String retorno = "[";
		int ponteiroAux = ponteiroInicio;
		for (int i = 0; i < quantidade; i++) {			
			if (i == quantidade - 1) 
				retorno += dados[ponteiroAux];
			else
				retorno += dados[ponteiroAux] + ",";
			
			ponteiroAux = avancar(ponteiroAux); 
		}
		return retorno + "]";		
	}

	//NOVO
	@Override
	public String imprimirDeTrasPraFrente() {
		String retorno = "[";
		int ponteiroAux = ponteiroFim;
		for (int i = 0; i < quantidade; i++) {			
			if (i == quantidade - 1) 
				retorno += dados[ponteiroAux];
			else
				retorno += dados[ponteiroAux] + ",";
			
			ponteiroAux = retroceder(ponteiroAux); 
		}
		return retorno + "]";		
	}
	
	private int avancar(int ponteiro) {
		return (ponteiro+1)%dados.length;
	}
	
	private int retroceder(int ponteiro) {
		return ((ponteiro-1)+dados.length)%dados.length;
	}	
}

File: fila/estatica/Kotlin/Fila Dupla Terminação/FilaEstaticaDuplaTerminacao.kt
class FilaEstaticaDuplaTerminacao(private val tamanho: Int = 10) : EnfileiravelDupla {
 
	private var ponteiroInicio = 0
	private var ponteiroFim = -1 
	private var dados: Array<Any?> = arrayOfNulls(tamanho)
    private var quantidade = 0

	override fun atualizarInicio(dado: Any?) {
		if (!estaVazia())
			dados[ponteiroInicio] = dado
		else
			println("Queue is empty!")
	}

	override fun atualizarFim(dado: Any?) {
		if (!estaVazia())
			dados[ponteiroFim] = dado
		else
			println("Queue is empty!")
	}

	override fun enfileirarFim(dado: Any?) {
		if (!estaCheia()) {
			ponteiroFim++
			//patch pra fila funcionar de forma circular
			if (ponteiroFim == dados.size)
				ponteiroFim = 0
			quantidade++
			//fim do patch			
			dados[ponteiroFim] = dado
		} else {
			println("Queue is full!")
		}
	}

    override fun enfileirarInicio(dado: Any?) {
        if (!estaCheia()) {
            ponteiroInicio--
            //patch pra fila funcionar de forma circular
            if (ponteiroInicio == -1) {
                ponteiroInicio = dados.size-1
                if (quantidade == 0)
                    ponteiroFim = dados.size-1
            }
            quantidade++
            //fim do patch			
            dados[ponteiroInicio] = dado
        } else {
            println("Queue is full!")
        }
    }

	override fun desenfileirarInicio(): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			dadoAux = dados[ponteiroInicio]
			ponteiroInicio++
			//patch pra fila funcionar de forma circular
			if (ponteiroInicio == dados.size)
				ponteiroInicio = 0
			quantidade--
			//fim do patch	
		} else {
			println("Queue is empty!")
		}
		return dadoAux
	}

	override fun desenfileirarFim(): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			dadoAux = dados[ponteiroFim]
			ponteiroFim--
			//patch pra fila funcionar de forma circular
			if (ponteiroFim == -1)
				ponteiroFim = dados.size-1
			quantidade--
			//fim do patch	
		} else {
			println("Queue is empty!")
		}
		return dadoAux
	}

	override fun frente(): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			dadoAux = dados[ponteiroInicio]
		} else {
			println("Queue is empty!")
		}
		return dadoAux
	}

	override fun tras(): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			dadoAux = dados[ponteiroFim]
		} else {
			println("Queue is empty!")
		}
		return dadoAux
	}
    
	override fun estaCheia(): Boolean {
		return quantidade == dados.size
	}
	
	override fun estaVazia(): Boolean {
		return quantidade == 0
	}
	
	override fun imprimirFrentePraTras(): String {
		var resultado = "["
		var ponteiroAux = ponteiroInicio
		for (i in 0 .. quantidade-1) {
			if (i == quantidade-1)
				resultado += "${dados[ponteiroAux]}"
			else
				resultado += "${dados[ponteiroAux]},"

			ponteiroAux++
            if (ponteiroAux == dados.size)
                ponteiroAux = 0			
		}
		return "$resultado]"
	}

	override fun imprimirTrasPraFrente(): String {
		var resultado = "["
		var ponteiroAux = ponteiroFim
		for (i in quantidade-1 downTo 0) {
			if (i == 0)
				resultado += "${dados[ponteiroAux]}"
			else
				resultado += "${dados[ponteiroAux]},"

			ponteiroAux--
            if (ponteiroAux == -1)
                ponteiroAux = dados.size-1			
		}
		return "$resultado]"
	}    
}